core/filterconstraint.cpp

```diff
@@ -0,0 +1,939 @@
+{
+	const string_mode_description *desc = get_string_mode_description(type);
+	return desc ? gettextFromC::tr(desc->translated) : "unknown";
+}
+
+QString filter_constraint_range_mode_to_string_translated(enum filter_constraint_range_mode type)
+{
+	const range_mode_description *desc = get_range_mode_description(type);
+	return desc ? gettextFromC::tr(desc->translated) : "unknown";
+}
+
+QString filter_constraint_negate_to_string_translated(bool negate)
+{
+	return gettextFromC::tr(negate_description[negate]);
+}
+
+static enum filter_constraint_units type_to_unit(enum filter_constraint_type type)
+{
+	const type_description *desc = get_type_description(type);
+	return desc ? desc->units : FILTER_CONSTRAINT_NO_UNIT;
+}
+
+QString filter_constraint_get_unit(enum filter_constraint_type type)
+{
+	switch (type_to_unit(type)) {
+	case FILTER_CONSTRAINT_NO_UNIT:
+	default:
+		return QString();
+	case FILTER_CONSTRAINT_LENGTH_UNIT:
+		return get_depth_unit();
+	case FILTER_CONSTRAINT_TEMPERATURE_UNIT:
+		return get_temp_unit();
+	case FILTER_CONSTRAINT_WEIGHT_UNIT:
+		return get_weight_unit();
+	case FILTER_CONSTRAINT_VOLUMETRIC_FLOW_UNIT:
+		return get_volume_unit();
+	}
+}
+
+static int float_to_int(double f, enum filter_constraint_type type)
+{
+	const type_description *desc = get_type_description(type);
+	if (!desc)
+		return -1;
+	switch (desc->units) {
+	case FILTER_CONSTRAINT_NO_UNIT:
+	default:
+		return (int)lrint(f);
+	case FILTER_CONSTRAINT_LENGTH_UNIT:
+		return prefs.units.length == units::METERS ? lrint(f * 1000.0) : feet_to_mm(f);
+	case FILTER_CONSTRAINT_TEMPERATURE_UNIT:
+		return prefs.units.temperature == units::CELSIUS ? C_to_mkelvin(f) : F_to_mkelvin(f);
+	case FILTER_CONSTRAINT_WEIGHT_UNIT:
+		return prefs.units.weight == units::KG ? lrint(f * 1000.0) : lbs_to_grams(f);
+	case FILTER_CONSTRAINT_VOLUMETRIC_FLOW_UNIT:
+		return prefs.units.volume == units::LITER ? lrint(f * 1000.0) : lrint(cuft_to_l(f) * 1000.0);
+	}
+}
+
+static double int_to_float(int i, enum filter_constraint_type type)
+{
+	const type_description *desc = get_type_description(type);
+	if (!desc)
+		return 0.0;
+	switch (desc->units) {
+	case FILTER_CONSTRAINT_NO_UNIT:
+	default:
+		return (double)i;
+	case FILTER_CONSTRAINT_LENGTH_UNIT:
+		return prefs.units.length == units::METERS ? (double)i / 1000.0 : mm_to_feet(i);
+	case FILTER_CONSTRAINT_TEMPERATURE_UNIT:
+		return prefs.units.temperature == units::CELSIUS ? mkelvin_to_C(i) : mkelvin_to_F(i);
+	case FILTER_CONSTRAINT_WEIGHT_UNIT:
+		return prefs.units.weight == units::KG ? (double)i / 1000.0 : grams_to_lbs(i);
+	case FILTER_CONSTRAINT_VOLUMETRIC_FLOW_UNIT:
+		return prefs.units.volume == units::LITER ? (double)i / 1000.0 : ml_to_cuft(i);
+	}
+}
+
+QStringList filter_constraint_type_list_translated()
+{
+	QStringList res;
+	for (size_t i = 0; i < type_descriptions_count(); ++i)
+		res.push_back(gettextFromC::tr(type_descriptions[i].translated));
+	return res;
```

Comment:
so here you are translating things marked as translated... I think you are setting yourself up for a ton of pain doing things this way. Maybe you answered that earlier, but I still don't understand why you would want to store the untranslated values, given that you already include a separate `token` (which you call `s`)

Reply:
As I said above - this is our common pattern for translated strings in the core. I renamed the member variable, which should make things more clear.